### PR TITLE
Improve loading time on Job API endpoint

### PR DESCRIPTION
**Pull request by vyzaldysanchez**
An attempt into improving the load time on the Jobs GQL endpoint

### CODE REVIEW

**Comment by vyzaldysanchez** [comment]
Would like a suggestion on how to improve this part... The `count` call might be making this endpoint slower. I understand it is needed for the pagination, but maybe we can only go once into the DB and get both the runs IDs for the Job and the total/count of runs in the same query?

@jmank88 @samsondav @jkongie

File: core/services/job/orm.go [hunk 1]
```diff
@@ -698,6 +701,42 @@ func (o *orm) FindJob(ctx context.Context, id int32) (jb Job, err error) {
...
+		if err != nil {
+			return errors.Wrap(err, "failed to load job")
+		}
+
+		if err = LoadAllJobTypes(tx, &jb); err != nil {
+			return errors.Wrap(err, "failed to load job types")
+		}
+
+		return nil
+	})
```
Comment (samsondav) [inline core/services/job/orm.go:722]: This could be a readonly transaction

**Review by samsondav — CHANGES_REQUESTED** [review]

File: core/services/job/orm.go (hunk 1 above)
Comment (vyzaldysanchez) [reply core/services/job/orm.go]: Nice catch, I wasn't aware of the existence of it! Thanks!

File: core/web/loader/job_spec_errors.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	// Collect the keys to search for
+	var jobIDs []int32
+	for ix, key := range keys {
+		id, err := stringutils.ToInt32(key.String())
+		if err == nil {
+			jobIDs = append(jobIDs, id)
+		}
+
+		keyOrder[key.String()] = ix
+	}
```
Comment (reductionista) [inline core/web/loader/job_spec_errors.go:29]: I notice we do the same song and dance for a lot of these batchers.  Seems like a good candidate for combing all of them into a more abstract batcher type using generics?

Just noting this as a suggestion for a future project, no need to include it in this PR.

File: core/web/loader/job_spec_errors.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+	specErrors, err := b.app.JobORM().FindSpecErrorsByJobIDs(jobIDs)
+	if err != nil {
+		return []*dataloader.Result{{Data: nil, Error: err}}
+	}
+
+	// Generate a map of spec errors to jobIDs
```
Comment (reductionista) [inline core/web/loader/job_spec_errors.go:36]: Looks like this should be "a map of jobIDs to spec errors"

**Comment by reductionista** [comment]
LGTM!

Left two comments, but it looks like both of them are issues which were already there in other files before this PR so nothing that should prevent merging.

**Review by reductionista — DISMISSED** [review]

File: core/web/loader/job_spec_errors.go (hunk 2 above)
Comment (samsondav) [reply core/web/loader/job_spec_errors.go:29]: Yes I think there could be a few generics-based loaders that would be handy

File: core/services/job/orm.go [hunk 4]
```diff
@@ -698,6 +701,42 @@ func (o *orm) FindJob(ctx context.Context, id int32) (jb Job, err error) {
 	return
 }
 
+// FindJobWithoutSpecErrors returns a job by ID, without loading Spec Errors preloaded
+func (o *orm) FindJobWithoutSpecErrors(id int32) (jb Job, err error) {
+	ctx, cancel := pg.DefaultQueryCtx()
+	defer cancel()
+
+	q := o.q.WithOpts(pg.WithParentCtx(ctx))
```
Comment (samsondav) [inline core/services/job/orm.go:709]: This context generation is unnecessary, `q` already encapsulates the default context

File: core/web/loader/job_spec_errors.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	"github.com/smartcontractkit/chainlink/core/services/chainlink"
+	"github.com/smartcontractkit/chainlink/core/services/job"
+	"github.com/smartcontractkit/chainlink/core/utils/stringutils"
+)
+
+type jobSpecErrorsBatcher struct {
+	app chainlink.Application
+}
+
+func (b *jobSpecErrorsBatcher) loadByJobIDs(_ context.Context, keys dataloader.Keys) []*dataloader.Result {
```
Comment (samsondav) [inline core/web/loader/job_spec_errors.go:17]: Please do not throw away the context

**Review by samsondav — CHANGES_REQUESTED** [review]

File: core/web/loader/getters.go [hunk 6]
```diff
@@ -232,3 +233,24 @@ func GetFeedsManagerChainConfigsByManagerID(ctx context.Context, mgrID int64) ([
...
+
+	thunk := ldr.SpecErrorsByJobIDLoader.Load(ctx,
+		dataloader.StringKey(stringutils.FromInt32(jobID)),
+	)
+	result, err := thunk()
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Printf("%v\n", result)
```
Comment (jmank88) [inline core/web/loader/getters.go:249]: Leftover?

File: core/web/loader/job_spec_errors.go (hunk 5 above)
Comment (vyzaldysanchez) [reply core/web/loader/job_spec_errors.go:17]: We are not really doing anything with the context here... should i pass it around to the ORM call and wrap that context as part of the whole process in the ORM?

File: core/web/loader/job_spec_errors.go (hunk 3 above)
Comment (vyzaldysanchez) [reply core/web/loader/job_spec_errors.go]: Addressed this comment

File: core/services/job/orm.go (hunk 4 above)
Comment (vyzaldysanchez) [reply core/services/job/orm.go]: Removed the context generation.

File: core/web/loader/getters.go (hunk 6 above)
Comment (vyzaldysanchez) [reply core/web/loader/getters.go]: Resolved.

File: core/web/loader/job_spec_errors.go (hunk 2 above)
Comment (vyzaldysanchez) [reply core/web/loader/job_spec_errors.go:29]: I will create a story to track this generics work around loaders.

**Review by reductionista — DISMISSED** [review]

**Review by jmank88 — APPROVED** [review]

**Review by samsondav — DISMISSED** [review]

**Review by jmank88 — APPROVED** [review]

**Pull request merged**